src/category_theory/with_terminal.lean

```diff
@@ -0,0 +1,366 @@
+def id : Π (X : with_terminal C), hom X X
+| (of X) := 𝟙 _
+| star := punit.star
+
+/-- Composition of morphisms for `with_terminal C`. -/
+@[simp]
+def comp : Π {X Y Z : with_terminal C}, hom X Y → hom Y Z → hom X Z
+| (of X) (of Y) (of Z) := λ f g, f ≫ g
+| (of X) _ star := λ f g, punit.star
+| star (of X) _ := λ f g, pempty.elim f
+| _ star (of Y) := λ f g, pempty.elim g
+| star star star := λ _ _, punit.star
+
+instance : category.{v} (with_terminal C) :=
+{ hom := λ X Y, hom X Y,
+  id := λ X, id _,
+  comp := λ X Y Z f g, comp f g }
+
+/-- The inclusion from `C` into `with_terminal C`. -/
+def incl : C ⥤ (with_terminal C) :=
+{ obj := of,
+  map := λ X Y f, f }
+
+instance : full (incl : C ⥤ _) :=
+{ preimage := λ X Y f, f }
+
+instance : faithful (incl : C ⥤ _) := {}
+
+/-- Map `with_terminal` with respect to a functor `F : C ⥤ D`. -/
+def map {D : Type*} [category D] (F : C ⥤ D) : with_terminal C ⥤ with_terminal D :=
+{ obj := λ X,
+    match X with
+    | of x := of $ F.obj x
+    | star := star
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := punit.star
+    | star, star, punit.star := punit.star
+    end }
+
+instance {X : with_terminal C} : unique (X ⟶ star) :=
+{ default :=
+    match X with
+    | of x := punit.star
+    | star := punit.star
+    end,
+  uniq := by tidy }
+
+/-- `with_terminal.star` is terminal. -/
+def star_terminal : limits.is_terminal (star : with_terminal C) :=
+limits.is_terminal.of_unique _
+
+/-- Lift a functor `F : C ⥤ D` to `with_term C ⥤ D`. -/
+@[simps]
+def lift {D : Type*} [category D] {Z : D} (F : C ⥤ D) (M : Π (x : C), F.obj x ⟶ Z)
+  (hM : ∀ (x y : C) (f : x ⟶ y), F.map f ≫ M y = M x) :
+  (with_terminal C) ⥤ D :=
+{ obj := λ X,
+    match X with
+    | of x := F.obj x
+    | star := Z
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := M x
+    | star, star, punit.star := 𝟙 Z
+    end }
+
+/-- The isomorphism between `incl ⋙ lift F _ _` with `F`. -/
+@[simps]
+def incl_lift {D : Type*} [category D] {Z : D} (F : C ⥤ D) (M : Π (x : C), F.obj x ⟶ Z)
+  (hM : ∀ (x y : C) (f : x ⟶ y), F.map f ≫ M y = M x) :
+  incl ⋙ lift F M hM ≅ F :=
+{ hom := { app := λ X, 𝟙 _ },
+  inv := { app := λ X, 𝟙 _ } }
+
+/-- The isomorphism between `(lift F _ _).obj with_terminal.star` with `Z`. -/
+@[simps]
+def lift_star {D : Type*} [category D] {Z : D} (F : C ⥤ D) (M : Π (x : C), F.obj x ⟶ Z)
+  (hM : ∀ (x y : C) (f : x ⟶ y), F.map f ≫ M y = M x) :
+  (lift F M hM).obj star ≅ Z := eq_to_iso rfl
+
+lemma lift_map_lift_star {D : Type*} [category D] {Z : D} (F : C ⥤ D) (M : Π (x : C), F.obj x ⟶ Z)
+  (hM : ∀ (x y : C) (f : x ⟶ y), F.map f ≫ M y = M x) (x : C) :
+  (lift F M hM).map (star_terminal.from (incl.obj x)) ≫ (lift_star F M hM).hom =
+  (incl_lift F M hM).hom.app x ≫ M x :=
+begin
+  erw [category.id_comp, category.comp_id],
+  refl,
+end
+
+/-- The uniqueness of `lift`. -/
+@[simps]
```

Comment:
```suggestion
@[simp]
```

Reply:
oops! thanks. (that was left over from before I was using `of_components`.)